### PR TITLE
Add crps losses to regressor

### PR DESCRIPTION
**Pull request by bejaeger**
This adds Ranked Probability Score (RPS) and Ranked Logarithmic Score (RLS) to the regressor finetuning code. 

Note that this PR builds upon the original work by Jonas Landsgesell and Pascal Knoll in #689. It adapts their implementation to be compatible with our latest finetuning refactor and adds RLS.

Adding the RPS loss in the finetune example results in significantly better performance.

To find good defaults for the regression loss weights, a broader benchmarking would need to be done.

### CODE REVIEW

**Comment by bejaeger** [comment]
This change is part of the following stack:

- #711 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: src/tabpfn/finetuning/finetuned_regressor.py [hunk 1]
```diff
@@ -30,41 +30,181 @@
...
+    target_cdf_BQL = (bin_indices_L.view(1, 1, -1) >= target_bins_BQ.unsqueeze(-1)).to(
+        probs_BQL.dtype
+    )
+
+    if loss_type == "rps":
+        cdf_diff_BQL = pred_cdf_BQL - target_cdf_BQL
+        cdf_term_losses_BQL = cdf_diff_BQL.square()
+    else:
+        eps = 1e-8
+        log_arg_BQL = (pred_cdf_BQL + target_cdf_BQL - 1.0).abs().clamp(min=eps)
```
Comment (noahho) [inline src/tabpfn/finetuning/finetuned_regressor.py]: RLS stability/clarity: the current -log(abs(CDF + y - 1)) form can be numerically unstable near the boundaries. Consider rewriting as the standard cumulative log score: y * -log(CDF) + (1-y) * -log(1-CDF) using log1p for stability:

eps = torch.finfo(pred_cdf_BQL.dtype).eps
cdf = pred_cdf_BQL.clamp(eps, 1 - eps)
cdf_term_losses_BQL =  target_cdf_BQL* (-torch.log(cdf)) + (1 - target_cdf_BQL) * (-torch.log1p(-cdf))

File: src/tabpfn/finetuning/finetuned_classifier.py [hunk 2]
```diff
@@ -29,7 +29,7 @@
     from tabpfn.finetuning.data_util import ClassifierBatch
 
 
-def compute_classification_loss(
+def _compute_classification_loss(
     *,
     predictions_BLQ: torch.Tensor,
```
Comment (noahho) [inline src/tabpfn/finetuning/finetuned_classifier.py]: Minor naming nit: in the regressor you switched to logits_BQL and the suffix order is BQL; here it’s still predictions_BLQ. Can we standardize suffix ordering + naming across classifier/regressor (e.g., logits_BQL everywhere)?

File: src/tabpfn/finetuning/finetuned_regressor.py [hunk 3]
```diff
@@ -30,41 +30,181 @@
...
+        total_loss = total_loss + rps_loss_weight * rps_loss
+
+    if rls_loss_weight > 0.0:
+        rls_loss = _ranked_probability_score_loss_from_bar_logits(
+            logits_BQL=logits_BQL,
+            targets_BQ=targets_BQ,
+            bardist_loss_fn=bardist_loss_fn,
+            loss_type="rls",
+        )
+        total_loss = total_loss + rls_loss_weight * rls_loss
```
Comment (noahho) [inline src/tabpfn/finetuning/finetuned_regressor.py]: Perf nit: _compute_regression_loss does multiple .mean() reductions and initializes total_loss via torch.tensor(0.0, device=...). Not a big deal, but we could avoid a new tensor alloc + extra reductions by accumulating per-element loss and doing a single .mean() at the end (or use logits_BQL.new_zeros(()) for the scalar).

Comment (noahho) [reply src/tabpfn/finetuning/finetuned_regressor.py]: Feel free to ignore

File: src/tabpfn/finetuning/finetuned_regressor.py (hunk 1 above)
Comment (noahho) [reply src/tabpfn/finetuning/finetuned_regressor.py]: Feel free to ignore if you don't think it's right, I'm not sure

Comment (bejaeger) [reply src/tabpfn/finetuning/finetuned_regressor.py]: Thanks, using `log1p` seems to be more numerically stable. Changed it.

File: src/tabpfn/finetuning/finetuned_classifier.py (hunk 2 above)
Comment (bejaeger) [reply src/tabpfn/finetuning/finetuned_classifier.py]: Good catch. I renamed to `logits` but kept the indices because we need to have the logits as second dimension for the CE loss. Wanted to avoid having 2 reshapes.

File: src/tabpfn/finetuning/finetuned_regressor.py (hunk 3 above)
Comment (bejaeger) [reply src/tabpfn/finetuning/finetuned_regressor.py]: I kept it the way it was cause I found it to be a bit more readable and performance doesn't seem important here.

File: src/tabpfn/finetuning/finetuned_regressor.py [hunk 4]
```diff
@@ -30,41 +30,185 @@
...
+
+    return total_loss
+
+
+def _ranked_probability_score_loss_from_bar_logits(
+    *,
+    logits_BQL: torch.Tensor,
+    targets_BQ: torch.Tensor,
+    bardist_loss_fn: Any,
+    loss_type: Literal["rps", "rls"] = "rps",
```
Comment (jonaslandsgesell) [inline src/tabpfn/finetuning/finetuned_regressor.py]:
```suggestion
    loss_type: Literal["crps", "crls"] = "crps",
```
I've updated the loss_type to CRPS and CRLS. In my previous draft, I overlooked a naming convention: when the weighting $w_k$ (the distance between bins) is explicitly included, we refer to it as the Continuous Ranked Probability Score. This is distinct from the standard RPS, which is often used for ordinal regression where the labels are ordered but the metric distance between them is undefined (e.g., treating the gap between labels 3 and 9999 the same as between 1 and 2). For our current implementation, keeping the weights is what we intend.

File: src/tabpfn/finetuning/finetuned_regressor.py [hunk 5]
```diff
@@ -131,10 +275,21 @@ class FinetunedTabPFNRegressor(FinetunedTabPFNBase, RegressorMixin):
 
         extra_regressor_kwargs: Additional keyword arguments to pass to the
             underlying `TabPFNRegressor`, such as `n_estimators`.
+        ce_loss_weight: Weight for the bar distribution negative log-likelihood term
+            (cross-entropy-like). Defaults to 0.0.
+        rps_loss_weight: Weight for the ranked probability score (RPS) term computed
```
Comment (jonaslandsgesell) [inline src/tabpfn/finetuning/finetuned_regressor.py]:
```suggestion
        crps_loss_weight: Weight for the continous ranked probability score (CRPS) term computed
```

File: src/tabpfn/finetuning/finetuned_regressor.py [hunk 6]
```diff
@@ -263,18 +428,23 @@ def _forward_with_loss(self, batch: RegressorBatch) -> torch.Tensor:  # type: ig
...
         )
 
-        return compute_regression_loss(
-            predictions_BLQ=predictions_BLQ,
+        return _compute_regression_loss(
+            logits_BQL=logits_BQL,
             targets_BQ=targets_BQ,
             bardist_loss_fn=bardist_loss_fn,
+            ce_loss_weight=self.ce_loss_weight,
+            rps_loss_weight=self.rps_loss_weight,
```
Comment (jonaslandsgesell) [inline src/tabpfn/finetuning/finetuned_regressor.py]:
```suggestion
            crps_loss_weight=self.crps_loss_weight,
```

File: src/tabpfn/finetuning/finetuned_regressor.py (hunk 4 above)
Comment (bejaeger) [reply src/tabpfn/finetuning/finetuned_regressor.py]: Thanks a lot for the clarification. I went ahead and used your suggested convention and changed all rps/rls -> crps/crls.

**Comment by jonaslandsgesell** [comment]
Great work Ben, thanks a lot for the integration in the current version. Really looking forward👍
